src/send/sender.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/warp-contracts/syncer/src/utils/bundlr/responses"
+	"github.com/warp-contracts/syncer/src/utils/config"
+	"github.com/warp-contracts/syncer/src/utils/model"
+	"github.com/warp-contracts/syncer/src/utils/monitoring"
+	"github.com/warp-contracts/syncer/src/utils/task"
+
+	"gorm.io/gorm"
+)
+
+type Sender struct {
+	*task.Task
+	rand    *rand.Rand
+	db      *gorm.DB
+	input   chan *model.DataItem
+	monitor monitoring.Monitor
+
+	// Bundling and signing
+	bundlrClient *bundlr.Client
+	signer       *bundlr.ArweaveSigner
+
+	// Updated data items
+	Output chan *model.DataItem
+}
+
+// Receives bundle items from the input channel and sends them to bundlr
+func NewSender(config *config.Config, db *gorm.DB) (self *Sender) {
+	var err error
+
+	self = new(Sender)
+	self.db = db
+
+	// Seed random generator
+	var b [8]byte
+	_, err = crypto_rand.Read(b[:])
+	if err != nil {
+		self.Log.WithError(err).Panic("Cannot seed math/rand package with cryptographically secure random number generator")
+	}
+	self.rand = rand.New(rand.NewSource(int64(binary.LittleEndian.Uint64(b[:]))))
+
+	self.Output = make(chan *model.DataItem)
+
+	self.Task = task.NewTask(config, "sender").
+		// Pool of workers that perform requests to bundlr.
+		// It's possible to run multiple requests in parallel.
+		// We're limiting the number of parallel requests with the number of workers.
+		WithWorkerPool(config.Sender.BundlerNumBundlingWorkers, config.Sender.WorkerPoolQueueSize).
+		WithSubtaskFunc(self.run)
+
+	self.signer, err = bundlr.NewArweaveSigner(config.Bundlr.Wallet)
+	if err != nil {
+		self.Log.WithError(err).Panic("Failed to create bundlr signer")
+	}
+
+	return
+}
+
+func (self *Sender) WithClient(client *bundlr.Client) *Sender {
+	self.bundlrClient = client
+	return self
+}
+
+func (self *Sender) WithInputChannel(in chan *model.DataItem) *Sender {
+	self.input = in
+	return self
+}
+
+func (self *Sender) WithMonitor(monitor monitoring.Monitor) *Sender {
+	self.monitor = monitor
+	return self
+}
+
+func (self *Sender) run() (err error) {
+	// Waits for new data items
+	// Finishes when when the source of items is closed
+	// It should be safe to assume all pending items are processed
+	for item := range self.input {
+		// Update stats
+		self.monitor.GetReport().Sender.State.AllBundlesFromDb.Inc()
+
+		// Copy the pointer so it's not overwritten in the next iteration
+		item := item
+
+		self.SubmitToWorker(func() {
+			if self.IsStopping.Load() {
+				// Don't start sending new items if we're stopping
+				return
+			}
+
+			var (
+				uploadResponse *responses.Upload
+				resp           *resty.Response
+				err            error
+			)
+
+			bundleItem, err := self.parse(item)
+			if err != nil {
+				self.Log.WithError(err).
+					WithField("data_item_id", item.DataItemID).
+					Error("Failed parse and validate data item")
+				item.State = model.BundleStateMalformed
+				goto end
+			}
+
+			// Send the bundle item to bundlr
```

Comment:
maybe we should rename all bundlr to iris?

Reply:
Done in sender